pkg/apiserver/certificate/certificate.go

```diff
@@ -178,15 +175,35 @@ func rotateSelfSignedCertificates(c *CACertController, secureServing *options.Se
 			klog.Errorf("error reading expiration date of cert: %v", err)
 			return
 		}
+
+		klog.Infof("Certificate will be rotated at %v", time.Now().Add(rotationDuration))
+
 		time.Sleep(rotationDuration)
 
 		klog.Infof("Rotating self signed certificate")
 
-		caContentProvider, err := generateSelfSignedCertificate(secureServing)
+		err = generateNewServingCertificate(secureServing)
 		if err != nil {
 			klog.Errorf("error generating new cert: %v", err)
 			return
 		}
-		c.UpdateCertificate(caContentProvider)
+		c.RunOnce()
```

Comment:
c.RunOnce calls `caContentProvider.RunOnce()` and `c.syncCACert()`,  while the former will trigger the latter too, then it will be executed twice.
I think it could just call `caContentProvider.RunOnce()`, the first argument of this function could be `*DynamicFileCAContent`.